onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -162,8 +265,48 @@ static Status InsertCastNodes(Graph& graph,
   return Status::OK();
 }
 
-// RemoveCastNodesChain
-// Remove the cast nodes specified in casts vector and fix the graph edges accordingly.
+/* RemoveCastNodesChain
+* Remove the cast nodes specified in casts vector and fix the graph edges accordingly. If the output of the cast
+* is also a graph output then insert an Identity node if the input of the cast node feeds other nodes. In the
```

Comment:
>Identity [](http://example.com/codeflow?start=40&length=8)

will it be eliminated in the final optimized training graph? if not, can we not to create Identity node?

Reply:
I think that the identity op will be removed because we run the optimizations in a loop multiple times and Cast propagation runs transformations in a loop each time until there are no changes.

Reply:
if your UTs cover this, then it should be fine, thx.

Reply:
The UT only runs the Propagate Cast transformation. I would expect that the Identity will removed further down the flow, if not at the graph transformation.